libraries/forms.py

```diff
@@ -501,6 +505,111 @@ def _get_library_versions(self, library_order, version):
+        if prior_version.release_date is None or version.release_date is None:
+            return None
+
+        @dataclass
+        class Day:
+            date: date
+            count: int
+            color: str = ""
+
+        @dataclass
+        class Week:
+            days: list[Day] = field(default_factory=list)
+
+            @cached_property
+            def max(self):
+                """The max number of commits this week."""
+                return max(x.count for x in self.days)
+
+        @dataclass
+        class Graph:
+            weeks: list[Week] = field(default_factory=list)
+            colors: list[str] = field(
+                default_factory=lambda: [
+                    "#E8F5E9",
+                    "#C8E6C9",
+                    "#A5D6A7",
+                    "#81C784",
+                    "#66BB6A",
+                    "#4CAF50",
+                    "#43A047",
+                    "#388E3C",
+                    "#2E7D32",
+                    "#1B5E20",
+                ],
+            )
+
+            @cached_property
+            def max(self):
+                """The max number of commits in all weeks."""
+                return max(x.max for x in self.weeks)
+
+            def append_day(self, day: Day):
+                """Append a day into the last week of self.weeks.
+
+                - Automatically create a new week if there are already 7 days in the
+                last week.
+                """
+                if len(self.weeks) == 0 or len(self.weeks[-1].days) == 7:
+                    self.weeks.append(Week())
+                self.weeks[-1].days.append(day)
+
+            def apply_colors(self):
+                """Iterate through each day and apply a color.
+
+                - The color is selected based on the number of commits made on
+                that day, relative to the highest number of commits in all days in
+                Graph.weeks.days.
+
+                """
+                high = self.max
+                for week in self.weeks:
+                    for day in week.days:
+                        decimal = day.count / high
+                        if decimal == 1:
+                            day.color = self.colors[-1]
+                        else:
+                            idx = int(decimal * len(self.colors))
+                            day.color = self.colors[idx]
+
+        count_query = (
+            Commit.objects.filter(library_version__version=version)
+            .annotate(date=TruncDate("committed_at"))
+            .values("date")
+            .annotate(count=Count("id"))
+        )
+        counts_by_date = {x["date"]: x["count"] for x in count_query}
+
+        graph = Graph()
+        # The start date is the release date of the previous version
+        # The end date is one day before the release date of the current version
+        start: date = prior_version.release_date
+        end: date = version.release_date - timedelta(days=1)
```

Comment:
It seems like we should filter the commits by date as well so we're not handling _all_ commits - just the ones we want to graph. Does this make sense? I haven't tested this.

Also, I don't think we need to use `TruncDate`. It's minimal, but I think there's a slight performance benefit to using the `__date` transform instead.

```suggestion
        # The start date is the release date of the previous version
        # The end date is one day before the release date of the current version
        start: date = prior_version.release_date
        end: date = version.release_date - timedelta(days=1)
        count_query = (
            Commit.objects.filter(
                library_version__version=version,
                committed_at__range=(start, end)),
            )
            .values("committed_at__date")
            .annotate(count=Count("id"))
        )
        counts_by_date = {x["committed_at__date"]: x["count"] for x in count_query}

        graph = Graph()
```

Reply:
> It seems like we should filter the commits by date as well so we're not handling all commits - just the ones we want to graph. Does this make sense?

Yes, that makes sense.  I think that querying for `library_version__version=version` is the way to do that filtering.  I think it's possible for commits which were created outside of the release date range to be merged into a future release, so the date filter may not be correct.  Does that sound right?

I'll use the __date transform.

Reply:
Oh yeah that makes sense. They should line up. 👍